src/domain/datasets/services/src/dataset_entry_service.rs

```diff
@@ -0,0 +1,239 @@
+    feeding_producers: &[
+        MESSAGE_PRODUCER_KAMU_CORE_DATASET_SERVICE,
+    ],
+    durability: MessageConsumptionDurability::Durable,
+})]
+#[interface(dyn InitOnStartup)]
+#[meta(InitOnStartupMeta {
+    job_name: JOB_KAMU_DATASETS_DATASET_ENTRY_INDEXER,
+    depends_on: &[JOB_KAMU_ACCOUNTS_PREDEFINED_ACCOUNTS_REGISTRATOR],
+    requires_transaction: true,
+})]
+pub struct DatasetEntryService {
+    dataset_entry_repo: Arc<dyn DatasetEntryRepository>,
+    time_source: Arc<dyn SystemTimeSource>,
+    dataset_repo: Arc<dyn DatasetRepository>,
+    account_repository: Arc<dyn AccountRepository>,
+}
+
+impl DatasetEntryService {
+    pub fn new(
+        dataset_entry_repo: Arc<dyn DatasetEntryRepository>,
+        time_source: Arc<dyn SystemTimeSource>,
+        dataset_repo: Arc<dyn DatasetRepository>,
+        account_repository: Arc<dyn AccountRepository>,
+    ) -> Self {
+        Self {
+            dataset_entry_repo,
+            time_source,
+            dataset_repo,
+            account_repository,
+        }
+    }
+
+    async fn handle_dataset_lifecycle_created_message(
+        &self,
+        message @ DatasetLifecycleMessageCreated {
+            dataset_id,
+            owner_account_id,
+            ..
+        }: &DatasetLifecycleMessageCreated,
+    ) -> Result<(), InternalError> {
+        let entry = DatasetEntry::new(
+            dataset_id.clone(),
+            owner_account_id.clone(),
+            message.dataset_name().into_owned(),
+            self.time_source.now(),
+        );
+
+        self.dataset_entry_repo
+            .save_dataset_entry(&entry)
+            .await
+            .int_err()
+    }
+
+    async fn handle_dataset_lifecycle_deleted_message(
+        &self,
+        DatasetLifecycleMessageDeleted { dataset_id, .. }: &DatasetLifecycleMessageDeleted,
+    ) -> Result<(), InternalError> {
+        self.dataset_entry_repo
+            .delete_dataset_entry(dataset_id)
+            .await
+            .int_err()
+    }
+
+    async fn handle_dataset_lifecycle_renamed_message(
+        &self,
+        DatasetLifecycleMessageRenamed {
+            dataset_id,
+            new_dataset_name,
+            ..
+        }: &DatasetLifecycleMessageRenamed,
+    ) -> Result<(), InternalError> {
+        self.dataset_entry_repo
+            .update_dataset_entry_name(dataset_id, new_dataset_name)
+            .await
+            .int_err()
+    }
+
+    async fn has_datasets_indexed(&self) -> Result<bool, InternalError> {
+        let stored_dataset_entries_count = self
+            .dataset_entry_repo
+            .dataset_entries_count()
+            .await
+            .int_err()?;
+
+        Ok(stored_dataset_entries_count > 0)
+    }
+
+    async fn index_datasets(&self) -> Result<(), InternalError> {
+        use futures::TryStreamExt;
+
+        let dataset_handles: Vec<_> = self.dataset_repo.get_all_datasets().try_collect().await?;
+
+        // TODO: in one transaction?
```

Comment:
It should be fine. If something fails, and we didn't save a thing to DB, it's easy to restart without garbage cleanup

Reply:
Transactional & concurrent processing added